core/client/db/src/cache/list_cache.rs

```diff
@@ -317,6 +320,44 @@ impl<Block: BlockT, T: CacheItemT, S: Storage<Block, T>> ListCache<Block, T, S>
 		Ok(Some(operation))
 	}
 
+	/// When block is reverted.
+	pub fn on_block_revert<Tx: StorageTransaction<Block, T>>(
+		&self,
+		tx: &mut Tx,
+		reverted_block: &ComplexBlockId<Block>,
+	) -> ClientResult<CommitOperation<Block, T>> {
+		// can't revert finalized blocks
+		debug_assert!(self.best_finalized_block.number < reverted_block.number);
+
+		// iterate all unfinalized forks and truncate/destroy if required
+		let mut updated = BTreeMap::new();
+		for (index, fork) in self.unfinalized.iter().enumerate() {
+			// we only need to truncate fork if its head is connected to truncated block
+			if !chain::is_connected_to_block(&self.storage, reverted_block, &fork.head.valid_from)? {
+				continue;
+			}
+
+			// we only need to truncate fork if its head is ancestor of truncated block
+			if fork.head.valid_from.number < reverted_block.number {
+				continue;
+			}
+
+			let updated_fork = fork.truncate(
+				&self.storage,
+				tx,
+				reverted_block.number,
+				self.best_finalized_block.number,
+			)?;
+			updated.insert(index, updated_fork);
```

Comment:
storing forks index means that the unfinalized map would not change before commit, I believe it is the case (if we combine multiple commitset pb could hapen (wonder if we can have a safer api by ensuring only one commitset is used but that is probably not very needed).

Reply:
I agree that the API could be better here. But since the cache is internal to the `substrate-client-db` and currently db crate only supports one operation per transaction (like import single block, revert blocks) let's leave this as is - it is faster than than anything else.